test/passing/break_cases_all.ml

```diff
@@ -4,80 +4,90 @@ let f x = function
   | C
   | P (this, test, [is; wide; enough; _to; break], [the; line])
   | A
-  | K ->
-      1
+   |K -> 1
   | D ->
       let a = "this" in
       let b = "breaks" in
       ()
 
 let f =
   let g = function
-    | H
-      when x y <> k ->
-        2
+    | H when x y <> k -> 2
     | T
-    | P
-    | U ->
-        3
+     |P
+     |U -> 3
   in
   fun x g t h y u ->
     match x with
-    | E ->
-        4
+    | E -> 4
     | Z
-    | P
-    | M -> (
+     |P
+     |M -> (
       match y with
-      | O ->
-          5
-      | P
-        when h x -> (
+      | O -> 5
+      | P when h x -> (
           function
-          | A ->
-              6 ) )
+          | A -> 6 ) )
 
 ;;
 match x with
 | true -> (
   match y with
-  | true ->
-      "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
-  | false ->
-      "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbb" )
-| false ->
-    "cccccccccccccccccccccccccccccc"
+  | true -> "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
+  | false -> "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbb" )
+| false -> "cccccccccccccccccccccccccccccc"
 
 ;;
 match x with
 | "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaa", yyyyyyyyyy
-  when fffffffffffffff bbbbbbbbbb yyyyyyyyyy ->
-    ()
-| _ ->
-    ()
+  when fffffffffffffff bbbbbbbbbb yyyyyyyyyy -> ()
+| _ -> ()
 
 let is_sequence exp =
   match exp.pexp_desc with
   | Pexp_sequence _
   | Pexp_extension
       ( _
       , PStr [{pstr_desc= Pstr_eval ({pexp_desc= Pexp_sequence _}, []); _}]
-      ) ->
-      true
-  | _ ->
-      false
+      ) -> true
```

Comment:
Just to check, is this output where a case body is placed on the same line at the end of a multi-line pattern intended? Technically, it is the same situation as after a nested or-pattern, but maybe not what is intended nevertheless.

Reply:
Also, I think a test case is needed where there is a multi-line pattern with a `when` clause, as in that case there needs to be a break before the `when`. There are a number of regressions of this form such as:
```
--- a/infer/src/IR/HilInstr.ml
+++ b/infer/src/IR/HilInstr.ml
@@ -68,8 +68,7 @@ let of_sil ~include_array_indexes ~f_resolve_id (instr : Sil.instr) =
       , Const (Cfun callee_pname)
       , (target_exp, _) :: (Sizeof {typ= cast_typ}, _) :: _
       , loc
-      , _ )
-    when Typ.Procname.equal callee_pname BuiltinDecl.__cast ->
+      , _ ) when Typ.Procname.equal callee_pname BuiltinDecl.__cast ->
       analyze_id_assignment (Var.of_id ret_id) target_exp cast_typ loc
   | Store (lhs_exp, typ, rhs_exp, loc) ->
       let lhs_access_expr =
```
I suspect that the change to avoid breaking before `when` for single-line patterns in `all` mode needs to be more narrowly targeted. But for multi-line patterns, I would argue that in all styles of match cases, there should be a break before`when` clauses on multi-or patterns. Without that, readers are strongly led to associate the `when` with the last pattern, which is wrong and a source of bugs.

Reply:
My last commit should fix that, running `test_branch` with `break-cases=all` on both looks like the output of this branch is more correct than it is on master.

Reply:
Looking quickly at the diff from `test_branch` with `break-cases=all`, I didn't spot any issues with `when`. But for the diff without setting any options, there are still regressions like the one in the diff above. This is in `nested` mode.

Using `break-cases=toplevel`, I still see situations where a case body follows a multiline pattern without a break after the `->`:
```
--- a/src/Ast.ml
+++ b/src/Ast.ml
@@ -1263,15 +1250,13 @@ end = struct
       | Pstr_primitive _ | Pstr_type _ | Pstr_typext _ | Pstr_exception _
        |Pstr_module _ | Pstr_recmodule _ | Pstr_modtype _ | Pstr_open _
        |Pstr_class _ | Pstr_class_type _ | Pstr_include _ | Pstr_attribute _
-        ->
-          assert false )
+        -> assert false )
     | Mod {pmod_desc= Pmod_unpack e1} -> (
       match e1 with
       | { pexp_desc=
             Pexp_constraint
               (e, {ptyp_desc= Ptyp_package _; ptyp_attributes= []})
-        ; pexp_attributes= [] } ->
-          assert (e == exp)
+        ; pexp_attributes= [] } -> assert (e == exp)
       | e -> assert (e == exp) )
     | Cl ctx ->
         let rec loop ctx =
```
Is this change intended, and intended to be combined with the case-breaking?

Reply:
These changes (fitting the rhs expression after the arrow if it fits even if the pattern broke) were intended, I reverted them for `all` and I can revert them for `nested` and `toplevel` as well.

Reply:
My impression is that the choice of whether to tack expressions on to the end of multiline patterns is mostly orthogonal to the style of breaking cases. Also, adding short phrases on the same line after multiline ones is a significant departure from the rest of the formatting. So I'd say that it would be best not to do it in the same PR.

In cases where the only thing at the beginning of the line is "punctuation" like closing parens or `->`, it would be fine. But the implementation of that could be complicated.

Reply:
And can I bump the request for a test case in 
https://github.com/ocaml-ppx/ocamlformat/pull/536#discussion_r252651181 . The situation with `when` seems subtle enough that it will help to have tests for it rather than only seeing such regressions on the full repo test diffs.